internal/controlplane/handlers_authz.go

```diff
@@ -344,85 +334,70 @@ func (s *Server) inviteUser(
+		Email:   email,
+		Project: targetProject,
+	})
 	if err != nil {
-		return nil, status.Errorf(codes.Unauthenticated, "failed to parse bearer token: %v", err)
+		return nil, status.Errorf(codes.Internal, "error getting invitations: %v", err)
 	}
 
-	// Get the sponsor's user information (current user)
-	currentUser, err := s.store.GetUserBySubject(ctx, openIdToken.Subject())
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to get user: %s", err)
+	// Check if there are any existing invitations for this email
+	if len(existingInvites) != 0 {
+		return nil, util.UserVisibleError(
+			codes.AlreadyExists,
+			"invitation for this email and project already exists, use update instead",
+		)
 	}
 
-	// Check if the user is already invited
-	existingInvite, err := s.store.GetInvitationByEmailAndProjectAndRole(ctx, db.GetInvitationByEmailAndProjectAndRoleParams{
+	// If there are no invitations for this email, great, we should create one
+	userInvite, err = s.store.CreateInvitation(ctx, db.CreateInvitationParams{
+		Code:    invite.GenerateCode(),
 		Email:   email,
-		Project: targetProject,
 		Role:    role.String(),
+		Project: targetProject,
+		Sponsor: currentUser.ID,
 	})
 	if err != nil {
-		if errors.Is(err, sql.ErrNoRows) {
-			// If there are no invitations for this email, great, we should create one
-			userInvite, err = s.store.CreateInvitation(ctx, db.CreateInvitationParams{
-				Code:    invite.GenerateCode(),
-				Email:   email,
-				Role:    role.String(),
-				Project: targetProject,
-				Sponsor: currentUser.ID,
-			})
-			if err != nil {
-				return nil, status.Errorf(codes.Internal, "error creating invitation: %v", err)
-			}
-		} else {
-			// Some other error happened, return it
-			return nil, status.Errorf(codes.Internal, "error getting invitations: %v", err)
-		}
-	} else {
-		// If we didn't get an error, this means there's an existing invite.
-		// We should update its expiration and send the response.
-		userInvite, err = s.store.UpdateInvitation(ctx, existingInvite.Code)
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "error updating invitation: %v", err)
-		}
+		return nil, status.Errorf(codes.Internal, "error creating invitation: %v", err)
 	}
-	// If we are here, this means we either created a new invite or updated an existing one
 
 	// Resolve the sponsor's identity and display name
-	identity := &auth.Identity{
-		Provider:  nil,
-		UserID:    currentUser.IdentitySubject,
-		HumanName: currentUser.IdentitySubject,
-	}
-	if flags.Bool(ctx, s.featureFlags, flags.IDPResolver) {
-		identity, err = s.idClient.Resolve(ctx, currentUser.IdentitySubject)
-		if err != nil {
-			zerolog.Ctx(ctx).Error().Err(err).Msg("error resolving identity")
-			return nil, util.UserVisibleError(codes.NotFound, "could not find identity %q", currentUser.IdentitySubject)
-		}
+	identity, err := s.idClient.Resolve(ctx, currentUser.IdentitySubject)
```

Comment:
We create an invitation then resolve the ID. What if resolving the ID fails for some reason? Should we instead wrap creating the invite and resolving the project ID into a transaction or just create the invite as the last step?

The concern is that if something fails and the invite is created, but the reply couldn't be sent, is it easy to delete this invite and start over?

Reply:
Addressed it in 1d54fdcfdf96692ac991451fba69eb8ac2c4a382

Reply:
thanks!